### ISSUE TITLE
Add minetest.check_tree to prevent crashes because of invalid treedefs [L-system]

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#4443]
Currently, if you feed an invalid tree definition into `minetest.spawn_tree`, Minetest crashes.
Okay, not really, but it stops the server and throws a Lua error. For server operators, this is equally bad.

One example for a bad tree definition would be for unbalanced brackets, like:

```
axiom="[[T]]]"
```

This bug report is not really new, in fact, I already reported it. It was “fixed” simply by shifting the responsibility to the modder. But I feel the underlying bigger issue here is still unresolved.

What I would like to see is that Minetest does not shift the entire responsibility to use valid treedefs to modders as this is more tricky than it looks on first sight.
For example, if a mod expects treedefs to come from user input, now the mod has to add a lot of boilerplate/checking code before actually calling `minetest.spawn_tree` as you can't trust user input, of course.
A concrete use case is the mod “L-System Tree Utility” [ltool] which is a tool to edit L-system trees in-game. One invalid tree entered into the form can crash the server if this tree is ever planted.

Why I think it is tricky to do the validity check:
Cases like “`axiom="[[T]]]"`” are simple enough, also anything which involves the letters A, B, C, and D is still realistic to solve; I just expand the tree definition along the route and then count all brackets.
But it gets tricky as soon any of the letters a-d is involved, since these are based on chance and I have multiple possible routes to follow and I have to check if ANY of them would result in unbalanced brackets, because Minetest may expand to any of those so any possible route with inbalanced brackets is a possible crash.
Only if all results have balanced brackets it would be fine. The thing is, if the check is only done in the tree generation itself, we have a very subtle kind of error, as the expansion of an invalid treedefs may or may not end up with unbalanced brackets, based on chance.
Example:

```
axiom="TTTa",
rules_a="]",
```

There is a 10% chance that this tree _doesn't_ crash the server.

Of course I could still try to solve it in my mod, but ultimately this should be in the engine, because if other mods like mine come along, they would have to add the exact same boilerplate code, which leads to redundancy. Better to fix it for all mods at once. Do you agree?

Here's how I think it should work:
I request to add the function `minetest.check_tree(treedef)` to be added. This function checks the validity of the supplied treedef, especially if any unbalanced brackets may appear in any way.
The function returns true for valid trees and false for invalid ones.

Normally, modders who have hardcoded their tree definition probably don't need to call this function before everytime they call `minetest.plant_tree` as they already probably know whether their definition is valid.

But for treedefs which come from untrusted sources calling this function will be neccessary.

**Comment by paramat** [comment]
We can't add code for checking mods for errors, that's the modder's responsibility, or should be done by an external program.

**Comment by Wuzzy2** [comment]
:-(
Did you even read the full post?
This case is special because it is not trivial to let users provide tree definitions and check them whether they're valid. Of course letting users provide trees needs some precautions but even privileged users should never be able to crash servers just by using malformed brackets. This is for my `ltool` mod.

Checking for obvious syntax errors like using an invalid character is trivial.

Checking if the brackets are always balanced is not. This matter is complicated by the fact that rules might not be executed (`a`, `b`, `c`, and `d` symbol).

External programs are useless here because it's about user input.
Of course I could fix it in my mod, but then similar mods would just repeat my code over and over again, leading to redundancy and eventually inconsistency.

**Comment by SmallJoker** [comment]
Minetest has to check the L-system params for validity and return `nil` if the placement failed, just like `minetest.place_schematic`.

**Comment by paramat** [comment]
> For example, if a mod expects treedefs to come from user input, now the mod has to add a lot of boilerplate/checking code before actually calling minetest.spawn_tree as you can't trust user input, of course.

Treedefs very rarely come from user input, so the mod has the responsibility to check the input.

> Of course I could still try to solve it in my mod, but ultimately this should be in the engine, because if other mods like mine come along, they would have to add the exact same boilerplate code, which leads to redundancy. Better to fix it for all mods at once. Do you agree?

No that's incorrect logic, if it's the responsibility of the mod then mods duplicating code is no issue.

However i agree with SmallJoker that an incorrect treedef should just result in no placement and a message in terminal without halting the server. So there is a valid request here, sorry for the close.
I disagree with analysing the treedef, but agree with better behaviour when the treedef fails.

**Comment by paramat** [comment]
As my penance i might work on this.

**Comment by Wuzzy2** [comment]
I have now employed successful workarounds in my `ltree` mod. Now when a player tries to save a treedef:

1. Check if axiom and all rulesets only contain valid character
2. Check for each axiom and all rulesets if they have balanced brackets

If any of these checks failed, the tree is rejected and cannot be saved by the player. These are trivial checks. This fixes the related crashes in my mod.
Note that my second rule restricts some valid treedefs, too, for example, when opening and closing brackets are mixed across rule sets. E.g.

A → [
B → F
C → ]
Axiom → ABC

is a valid L-System tree because it resolves to “[F]”, but it would be rejected under rule 2, because A and C have unbalanced brackets *individually*.

I agree, deep alalysis of the treedefs would be too complex.
Unbalanced brackets can sometimes be really nasty in that they only sometimes crash if you use the letters a-d, because those are only resolved by random chance. 

This is a Russian-Roulette tree: :grinning:
D → [
A → FFFF+FFFF+FFFF+FFFF
Axiom → Ad

I have thought a while about this problem. The lame solution is to simply resolve the treedef to all possible paths for each lower-case a-d. But this ultra-lame because the amount of paths you have to resolve rises exponentially. I have tried to figure out whether a more efficient solution is possible, but I decided it's not actually worth the hassle.

I think the rule 2 is perfectly reasonable and the loss of flexibility is acceptable. I have never seen a treedef which spreads brackets across rule sets and this use case seems rather contrived. Please correct me if I'm wrong.
Note this is just what I did for my mod, I'm note sure if it should be added as-is into Minetest.


Long story short, I do not really need the requested engine check anymore, but I still think it's better to fail gracefully.

**Comment by erlehmann** [comment]
>  I have tried to figure out whether a more efficient solution is possible, but I decided it's not actually worth the hassle.

I suspect this boils down to the question “Is the given grammar more complex than [deterministic] context-free?”:
https://en.wikipedia.org/wiki/L-system#L-system_structure

@appgurueu do you have experience with L-systems maybe?

**Comment by appgurueu** [comment]
Your description says a "Lua Error" is thrown. Why can't you catch that using `[x]pcall` and fail gracefully?

In general, there are two Lua patterns for error handling:

1. `return nil, errmsg` in case of an error. Can be used with `assert` as `local val = assert(func(...))` to assert that the call has succeeded. This is useful if errors are frequently expected and should thus be "silenced" to be handled conveniently as nil values by program logic. `tonumber` f.E. does this.
2. Throw the error using `error(errmsg)`. Can be catched with `[x]pcall`. This is for rather "rare" errors that *should* bubble all the way up. I'd argue that this is the correct behavior here: The user has supplied an incorrect tree which is must throw an error. Placing L-system trees is usually not exposed to users. Wuzzy's mod is the exception, not the norm, as it's a dev tool making devs users.

**Comment by appgurueu** [comment]
Handling `abcd` is actually rather easy: Get rid of them and check whether the tree def is valid. If it isn't, `abcd` don't change that in (100%) of cases, since it's possible that all of them don't get substituted. If the tree def is valid but any of `abcd` might evaluate to unbalanced parentheses, the tree isn invalid, since exactly one substitution is possible.

The algorithm would be roughly as follows:

1. Check for invalid characters. Remove all characters except for `[]AaBbCcDd`.
2. The only remaining possible issue are unbalanced parentheses *after all substitutions have been made*. The simplest way to do this is what Minetest already does: Plant the tree, evaluate it, check whether you encounter an error. This unfortunately doesn't work when `abcd` are involved, since they might "save" an invalid tree *sometimes*, but not *always*, and `check_tree` should return whether a tree is *always* valid.
